[Chiaryan] iP

chiaryan opened the pull request:
> meow meow meow meow meow meow – Louie when he wants food
# Intro...
#
#
#
#
#
#
#
#
#
# ...ducing...
#
#
#
#
#
#
# Louie!!!!!!

# a heading
- a 
- bullet 
- list

1. a
2. numbered
3. list
```java
// a fenced code block (with syntax highlighting)
this.print("HI!!!!!!!!!!!!! ");
```

- [x] a 
- [ ] task
- [ ] list

an emoji 🌚 

> a blockquote
[a hyperlink](https://i.nuuls.com/x_tf4.jpeg)
`inline code`
some text formatting: **bold**, _italic_, ~strikethrough~ etc.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

chiaryan committed:
Changed name and aesthetic
Added custom printing for chatbot output

chiaryan committed:
changed text-ui-test for new behaviour

chiaryan committed:
Level-1 requirement: echo user input

chiaryan committed:
add classes for Commands and Tasks

chiaryan committed:
Implement Task marking & unmarking

chiaryan committed:
Level-2 requirement: task listing

chiaryan committed:
Level-3 requirement: marking/unmarking tasks
Refactor: move task list and command mapping to main class

chiaryan committed:
ToDo, Deadline and Event classes

chiaryan committed:
Level-4 requirement: ToDo, Deadline and Event tasks

chiaryan committed:
Adding Excpetions

chiaryan committed:
Update the list, bye and mark unmark commands to use exceptions

chiaryan committed:
Update deadline command to use exceptions

chiaryan committed:
Update event command to use exceptions

chiaryan committed:
Add test for the previous commit

chiaryan committed:
Level-6 requirement: Implement deleting of tasks from list

chiaryan committed:
Change the Task class to abstract

chiaryan committed:
Add implementation for task list storage

chiaryan committed:
Correct storage string for tasks.

chiaryan committed:
Implement task storing for relevant commands

chiaryan committed:
Add format checking when reading from storage file

chiaryan committed:
ignore text-ui-text\data.txt

chiaryan committed:
Level-8 Requirement: Format task deadline/start and end dates to actually use LocalDateTimes

chiaryan committed:
Merge branch 'branch-Level-7'

chiaryan committed:
Merge branch 'branch-Level-8'

chiaryan committed:
OOP: Encapsulate Ui as an object

chiaryan committed:
Update tests for date-time strings

chiaryan committed:
OOP: Encapsulate the list of commands and querying actions on them

chiaryan committed:
Replace a forgotten new Ui().print()

chiaryan committed:
OOP: Encapsulate the list of tasks as a class TaskList

chiaryan committed:
OOP: Added a Parser class to process user input

chiaryan committed:
Organise all classes into a package

chiaryan committed:
Fix runtest.bat to properly compile the java files in the duke folder

chiaryan committed:
Fixed io test for new error reporting

chiaryan committed:
Merge branch 'add-gradle-support'

chiaryan committed:
Add gradle support

chiaryan committed:
Add tests for the ToDo class (describe and toStorageString)

chiaryan committed:
TodoTest: Swap assertEquals parameters to match expected/actual

chiaryan committed:
Add tests for Deadline class

chiaryan committed:
Add test for out-of-bound time in Deadline constructor

chiaryan committed:
Add TaskTest.java

The Task.fromStorageStrings method has no tests for
correct/incorrect inputs

Let's add some in a new test class

the tests include:
* correct inputs that create each of the Task subclasses
* erroneous cases where the input
  * has too many/too little parameters
  * has parameters that are empty strings
  * contains invalid dates
  * contains invalid values for the task type/completed
  * has incorrect parameters for the task type

Currently the tests pass inputs with too many parameters,
this is a bug that should be fixed in the near future...

chiaryan committed:
Add find command

This is part of the requirement Level-9: Find
for the iP.

Add a command 'find' to Duke, which prints all tasks
whose name contains the given substring

Tests for the relevant methods yet to be created.

chiaryan committed:
Add tests for Task.filterSubstring

There are no tests available for the above
method

Lets add some tests to check for queries that
will retrieve:
* all the tasks
* one task
* a subset of tasks
* no tasks at all

chiaryan committed:
Add JavaDoc for the Command class

chiaryan committed:
Add JavaDoc for the CommandList class

chiaryan committed:
Add JavaDoc for the Deadline class

chiaryan committed:
Add JavaDoc for Exceptions

chiaryan committed:
Add JavaDoc for TaskList

chiaryan committed:
Add JavaDoc for Parser

chiaryan committed:
Add declaration of AI usage

chiaryan committed:
Add MANIFEST.MF whatever

chiaryan committed:
Adjust code to fit the Java coding standard.

This is to fulfil the A-CodingStandard
increment requirement.

chiaryan committed:
Merge branch 'branch-A-CodingStandard'

chiaryan committed:
Merge branch 'branch-A-JavaDoc'

chiaryan committed:
Merge branch 'branch-Level-9'

shamesjen commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,334 @@
+    
+    private static final Ui ui = new Ui();
+
+    private static boolean done = false;
```
should this boolean be named isDone?

chiaryan replied:
It should 😄. It should be fixed now.

shamesjen commented on src/main/java/duke/DukeCommandNotFoundException.java:
```diff
@@ -0,0 +1,7 @@
+package duke;
```
I like the way you have different exceptions to treat different cases

shamesjen reviewed (commented):
Great Work! Easy to read code with different exceptions for different scenarios.

choowengyan commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,334 @@
+    
+    private static TaskList tasks = new TaskList();
+
+    private static final Storage st = new Storage("data.txt");
```
Should this constant be upper case?

choowengyan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,85 @@
+     * @throws DukeOptionParsingException If all tokens have been read from the input
+     */
+    public String next() throws DukeOptionParsingException {
+        String ret = this.peek();
```
Perhaps a more intuitive variable name here?

chiaryan replied:
I think the name is fine for now given the size of the code block, though I'll rename it to `token` for readability, in case the complexity of the code increases. Thanks!

choowengyan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,85 @@
+
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+/**
```
I like how you provided your comments to each of your methods here!

choowengyan commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,107 @@
+        }
+        doneStr = sc.next();
+        
+        switch (typeStr) {
```
I like how you follow the SE guidelines for switch case indentation!

choowengyan commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,99 @@
+ */
+public class TaskList {
+
+    private final ArrayList<Task> list = new ArrayList<>();
```
Perhaps would be better if the variable name here be `lists`?

chiaryan replied:
Noted on the naming violation. Will rename to `tasks` in the next commit.

choowengyan reviewed (commented):
Overall looks good and codes are well-structured!

chiaryan committed:
Add checkstyle files.

These files contain the code styling rules for the course's
coding standard.

chiaryan committed:
Configure build.gradle to import checkstyle

This will allow the project to check whether it aligns with
the course's coding standard.

Actual changes to the code are yet to be done.

chiaryan committed:
Implement Basic Gui

This is a milestone towards a full GUI chatbot. currently it
can echo user input

chiaryan committed:
Implement user input reading.

This is a milestone towards a full GUI chatbot. Duke can now
interpret user input (though it still speaks through stdout)

chiaryan committed:
Implement Chatbot responding.

The chatbot could only respond in stdin.

Lets change it so that the chatbot can use the GUI to speak
like the user

chiaryan committed:
Fix application exit command

The old 'bye' command changes an internal variable which
allows the chatbot exit a loop. This doesn't work in JavaFX's
event-driven system

This change uses the Platform.exit() method to exit the
application

chiaryan committed:
Rename variables to adhere to coding convention.

This is done based on the feedback on the PR.

chiaryan committed:
Reorganise code into packages.

chiaryan committed:
Separate all command logic into separate classes.

Command logic used to be added as lambdas during runtime
which resulted in a very long main method.

Lets separate all the commands as subclasses of
Command.

chiaryan committed:
Separate out UI code

The entire code for the rendering of the application UI is
all inside the Duke start method.

Let's move this code to the Ui class to keep the chatbot
logic and UI separate.

chiaryan committed:
Add parser utilities

chiaryan committed:
Shorten Task.fromStorageString method

The method uses alot of repetition such as for checking end
of input.

Let's condense the method by handling premature end of input
using exception handling

chiaryan committed:
Refactor out response strings

The messages/String.format templates are literals passed
to the respective print functions.

Let's define these strings as named constants to improve the
readability of the code.

chiaryan committed:
Add basic assertions

This is to complete the A-Assertions requirement. the iP has
no assertions in place to verify basic programming
assumptions

Let's add some basic assertions for the following conditions:
- Arguments passed to methods should not be null.
- Task names should never contain newlines.
- Some assumptions about program state after parsing.

chiaryan committed:
Add priority tagging for tasks.

This is to fulfil the BCD-Extension requirement

This change gives the ability to prioritise, deprioritise,
and view prioritised tasks.

chiaryan committed:
Bugfix: Parser.rest does not advance the cursor to the end.

chiaryan committed:
Break up fromStorageString method.

The fromStorageString method is too long.

Lets separate out some of the logic that reads and validates
the storage strings.

chiaryan committed:
Modify Tests for Priority.

Changing the bot for priority means that earlier unit tests
for displaying tasks will become invalidated.

Let's bring the tests up to date to include priority tagging.

chiaryan committed:
Add UI screenshot

chiaryan committed:
Rename 'Duke' to Louie

Let's create a personality for Louie by changing all
instances of the name 'Duke'.

chiaryan committed:
Create User Guide.

The iP lacks a User Guide to show how to use the chatbot
through commands.

chiaryan committed:
Update build-gradle for creating fat jar file

chiaryan committed:
Amendment for Java 11.

The iP need to be built for Java 11 for iP grading.

Some syntax changes are required for compatibility

chiaryan committed:
Update user guide to include download instructions

chiaryan committed:
Create assertion for toStorageString method

The name of a task can never contain a delimiter since it checks
during construction. Lets add an assetion for when the tasks are
being stored.

chiaryan committed:
create constant for name not containing commas error

This is to prevent repetition of message strings.

chiaryan committed:
Merge pull request #2 from chiaryan/branch-A-Assertions

Create assertion for toStorageString method

chiaryan committed:
Merge pull request #3 from chiaryan/branch-ACodeQuality

create constant for name not containing commas error

chiaryan committed:
Update docs with new release URL

With the new v0.3 release, The URL in the README.md points to the old vo.2 release.

chiaryan committed:
Add link to installation guide to product website

Users landing on the product webpage will have no idea how to
download the JAR file/run the bot.

Lat's add a link to the installation guide on the README.md.

chiaryan committed:
Update Ui screenshot

The current Ui.png has the text box cropped out.

Let's add a more complete screenshot.
